components/places/src/api/placesapi.rs

```diff
@@ -0,0 +1,256 @@
+        let db_name = db_name.into().canonicalize()?;
+        Self::new_or_existing(db_name, encryption_key)
+    }
+
+    /// Create a new, or fetch an already open, memory-based PlacesApi. You must
+    /// provide a name, but you are still able to have a single writer and many
+    ///  reader connections to the same memory DB open.
+    pub fn new_memory(db_name: &str, encryption_key: Option<&str>) -> Result<Arc<Self>> {
+        let name = PathBuf::from(format!("file:{}?mode=memory&cache=shared", db_name));
+        Self::new_or_existing(name, encryption_key)
+    }
+
+    fn new_or_existing(db_name: PathBuf, encryption_key: Option<&str>) -> Result<Arc<Self>> {
+        // XXX - we should check encryption_key via the HashMap here too.
+        let mut guard = APIS.lock().unwrap();
+        match guard.get(&db_name) {
+            Some(existing) => Ok(existing.clone()),
+            None => {
+                // We always create a new read-write connection for an initial open so
+                // we can create the schema and/or do version upgrades.
+                let connection =
+                    PlacesDb::open(&db_name, encryption_key, ConnectionType::ReadWrite)?;
+                let new = PlacesApi {
+                    db_name: db_name.clone(),
+                    encryption_key: encryption_key.map(|x| x.to_string()),
+                    write_connection: Mutex::new(Some(connection)),
+                    sync_state: Mutex::new(None),
+                };
+                let arc = Arc::new(new);
+                (*guard).insert(db_name, arc.clone());
+                Ok(arc)
+            }
+        }
+    }
+
+    /// Open a connection to the database.
+    pub fn open_connection(&self, conn_type: ConnectionType) -> Result<PlacesDb> {
+        let ec = self.encryption_key.as_ref().map(|x| x.as_str());
+        match conn_type {
+            ConnectionType::ReadOnly => {
+                // make a new one - we can have as many of these as we want.
+                PlacesDb::open(self.db_name.clone(), ec, ConnectionType::ReadOnly)
+            }
+            ConnectionType::ReadWrite => {
+                // We only allow one of these.
+                let mut guard = self.write_connection.lock().unwrap();
+                match mem::replace(&mut *guard, None) {
+                    None => Err(ErrorKind::ConnectionAlreadyOpen.into()),
+                    Some(db) => Ok(db),
+                }
+            }
+            ConnectionType::Sync => {
+                // ideally we'd enforce this in the same way as write_connection
+                PlacesDb::open(self.db_name.clone(), ec, ConnectionType::Sync)
+            }
+        }
+    }
+
+    /// Close a connection to the database. If the connection is the write
+    /// connection, you can re-fetch it using open_connection.
+    pub fn close_connection(&self, connection: PlacesDb) -> Result<()> {
```

Comment:
If `connection` is a connection from a different PlacesApi this will subtly break.

In general we, at a minimum, probably want to keep a global atomic counter for PlaceApi ids, store it on the Connection, and complain if the id doesn't match.

However, I suspect that the work we'll want to do to firm up how some of this works might make that unnecessary.